src/app/components/shared/input/date/date.component.ts

```diff
@@ -0,0 +1,310 @@
+      this.minDate = convertStructToDate(this.minDateStruct);
+    }
+
+    if (this.maxDate) {
+      this.maxDateStruct = convertDateToStruct(this.maxDate);
+    } else {
+      this.maxDate = convertStructToDate(this.maxDateStruct);
+    }
+  }
+
+  /** Invoked when the model has been changed */
+  public onChange: (_: Date) => void = () => {};
+
+  /** Invoked when the model has been touched */
+  public onTouched: () => void = () => {};
+
+  /** Method that is invoked on an update of a model. */
+  public updateChanges = () => this.onChange(this.value);
+
+  /**
+   * Registers a callback function that should be called when the control's value changes in the UI.
+   *
+   * @param fn
+   */
+  public registerOnChange = (fn: any): void => (this.onChange = fn);
+
+  /**
+   * Registers a callback function that should be called when the control receives a blur event.
+   *
+   * @param fn
+   */
+  public registerOnTouched = (fn: any): void => (this.onTouched = fn);
+
+  /**
+   * Method that is invoked when the control status changes to or from "DISABLED".
+   */
+  public setDisabledState = (isDisabled: boolean) =>
+    (this.disabled = isDisabled);
+
+  /**
+   * Writes a new item to the element.
+   *
+   * @param value the value
+   */
+  public writeValue(value: string): void {
+    this._errors = this.validateInput(value);
+    this.setDate(this.normalizeInput(value));
+    if (this.errors.length === 0) {
+      this.onChange(this.value);
+    }
+  }
+
+  /**
+   * Method that performs synchronous validation against the provided control
+   */
+  public validate(control: AbstractControl): ValidationErrors | null {
+    return this.validateInput(control.value);
+  }
+
+  public get showValid(): boolean {
+    return isInstantiated(this.value) && this.touched && !this.showError;
+  }
+
+  public get showError(): boolean {
+    return shouldShowError(this.dirty, this.touched, this.errors);
+  }
+
+  public get errors(): string[] {
+    return getErrorMessages(this._errors, this.errorTypes);
+  }
+
+  public validateInput(value: string | null): ValidationErrors | null {
+    if (!value) {
+      // No value provided on required input
+      return this.required ? { required: true } : null;
+    } else if (value.length < this.inputLength) {
+      // Validate length of input
+      return { minLength: true };
+    } else if (value.split("-").length !== 3) {
+      // Validate delimiters exist
+      return { delimiter: true };
+    }
+
+    const segments = value.split("-");
+    const year = parseInt(segments[0], 10);
+    const month = parseInt(segments[1], 10);
+    const day = parseInt(segments[2], 10);
+
+    if (year < this.minDateStruct.year || year > this.maxDateStruct.year) {
+      // Validate year
+      return { year: true };
+    } else if (month < 1 || month > 12) {
+      // Validate month
+      return { month: true };
+    } else if (day < 1 || day > 31) {
+      // Validate day
+      return { day: true };
+    }
```

Comment:
Why are checking only the year component is within range?

Reply:
It is far more complex to validate the month and day ranges are within range, as it only matters if the year matches the min or max dates. Much simpler to handle this process after conversion to dates is done

Reply:
Huh?

```
x = Date.parse(...)
if (x < this.minDate || x > this.maxDate) {
  // fail validation?
}
```

Why is it harder?

Your component provides this definition:

```
  /** Minimum date allowed in selection */
  @Input() public minDate: Date;
  /** Maximum date allowed in selection */
  @Input() public maxDate: Date;
```

It in no way suggests that only the year component from an entire date object is considered in validating it is in range.

Reply:
Oh I see. You validate for if the date is in range below. In which case this year validation is simply redundant and not needed.